src/main/java/de/tum/in/www1/artemis/service/connectors/BambooService.java

```diff
@@ -429,78 +427,98 @@ public String getPlanKey(Object requestBody) throws BambooException {
         }
     }
 
+    /**
+     * React to a new build result from Bamboo, create the result and feedbacks and link the result to the submission and participation.
+     *
+     * @param participation The participation for which the build finished.
+     * @param requestBody   The request Body received from the CI-Server.
+     * @return the created result.
+     * @throws Exception
+     */
     @Override
+    @Transactional
     public Result onBuildCompletedNew(ProgrammingExerciseParticipation participation, Object requestBody) throws Exception {
         log.debug("Retrieving build result (NEW) ...");
         try {
             Map<String, Object> requestBodyMap = (Map<String, Object>) requestBody;
             Map<String, Object> buildMap = (Map<String, Object>) requestBodyMap.get("build");
-            String buildReason = (String) buildMap.get("reason");
-            if (buildReason != null && buildReason.contains("First build for this plan")) {
-                //Filter the first build plan that was automatically executed when the build plan was created
-                return null;
-            }
-
-            Result result = new Result();
-            result.setRatedIfNotExceeded(participation.getProgrammingExercise().getDueDate(), ZonedDateTime.now());
-            result.setAssessmentType(AssessmentType.AUTOMATIC);
-            result.setSuccessful((Boolean) buildMap.get("successful"));
 
-            Map<String, Object> testSummary = (Map<String, Object>) buildMap.get("testSummary");
-            result.setResultString((String) testSummary.get("description"));
+            // Filter the first build plan that was automatically executed when the build plan was created.
+            if (isFirstBuildForThisPlan(buildMap)) return null;
 
-            result.setCompletionDate(ZonedDateTime.parse((String) buildMap.get("buildCompletedDate")));
-            result.setScore(calculateScoreForResult(result));
-            result.setBuildArtifact((Boolean) buildMap.get("artifact"));
-            result.setParticipation((Participation) participation);
+            // TODO: This should only fetch the submissions that don't have a result (yet).
+            List<ProgrammingSubmission> submissions = programmingSubmissionRepository.findByParticipationIdOrderByIdDesc(participation.getId());
+            Optional<ProgrammingSubmission> latestMatchingPendingSubmission = submissions.stream().filter(s -> s.getResult() == null).filter(s -> {
+                String matchingCommitHashInBuildMap = getCommitHash(buildMap, s.getType());
+                return matchingCommitHashInBuildMap.equals(s.getCommitHash());
+            }).findFirst();
```

Comment:
In the case of a test repository commit, it is not possible to use the provided commit hash to identify the submission:
- When the VCS informs Artemis about a new test repository commit, a submission is created for every participation with the commitHash **of the test repository**
- When the CI informs Artemis about a new build result, it is now unclear if the VCS last commit can be used to match the submission as the submission's commitHash might both be from the test repository OR the assignment repository

Reply:
What I'm doing here is to both match pending submissions AND the commitHash to identify the correct commitHash by type:
- submission is pending and has type `TEST`? => We need to check the test repository commit hash
- submission is pending and has type `MANUAL`? => We need to check the assignment repository commit hash

Reply:
Your approach seems good for me, so am I right that the issue with detecting the correct submission is solved now (as long as Artemis gets informed of the test-case update **before** the first build results arrive)?